Lazy Methods not loading base classes of params

thk123 opened the pull request:
Fixes a bug when using lazy methods would not load the base classes methods if the class was only referenced a field of another loaded class.

I have also pulled out the context insensitive lazy methods code (a collection of static functions) into its own class. (Second commit: https://github.com/diffblue/cbmc/pull/1228/commits/bb6184ba8307a711a2a7eaf8d78c0608258af180) let me know if you don't think it is an improvement. 

You will definitely want to review this commit by commit!

reuk commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -340,32 +346,62 @@ static void gather_needed_globals(
+  std::vector<irep_idt> class_and_parents=
+    ch.get_parents_trans(param_classid);
+
+  class_and_parents.push_back(param_classid);
```
I would prefer to write `lazy_methods.add_needed_class(param_classid)` here instead. At the moment, calling `push_back` here may cause a reallocation which can be avoided.

thk123 replied:
Change applied but thanks to @smowton suggestion no longer required.

smowton commented on src/java_bytecode/gather_methods_lazily.cpp:
```diff
@@ -0,0 +1,499 @@
+///   we should assume are reachable
+/// \param ns: global namespace
+/// \param ch: global class hierarchy
+/// \param pointer_type_selector: The pointer_type_selector used to find out
```
Parameter doesn't exist (yet?)

thk123 replied:
Good spot (yes, yet :P)

smowton commented on src/java_bytecode/gather_methods_lazily.cpp:
```diff
@@ -0,0 +1,499 @@
+#include <java_bytecode/java_string_library_preprocess.h>
+
+
+gather_methods_lazilyt::gather_methods_lazilyt(
```
Suggest keeping the `ci_lazy_methods` name (CI, context-insensitive, will differentiate this from future smarter ways of lazily faulting methods in). Alternatively, rename e.g. the constants in `java_bytecode_languaget` and the documentation to be consistent with any renaming.

thk123 replied:
Yup good idea - will do

smowton commented:
Aha, and here's a correct implementation of the tree walk, keeping track of the current override as we walk the tree: https://github.com/diffblue/cbmc/blob/test-gen-support/src/goto-programs/remove_virtual_functions.cpp#L255

thk123 commented:
@smowton ah missed your follow up comment - I applied a basic fix that walks up the tree when method implementations are missing but will see if I can hook in that analysis instead.

thk123 commented:
@smowton I think this is ready for re-review. You may wish to exclude the final commit that renames the files to make the diff more reasonable.

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -258,32 +260,24 @@ void remove_virtual_functionst::get_functions(
+
+  resolve_concrete_function_callt call_resolver(symbol_table);
+  const resolve_concrete_function_callt ::concrete_function_callt &
+    resolved_call=call_resolver(class_id, component_name);
```
`call_resolver` is perhaps a little vauge, how about `get_virtual_call_target`?

smowton commented on src/goto-programs/resolve_concrete_function_call.cpp:
```diff
@@ -0,0 +1,87 @@
+/*******************************************************************\
+
+ Module: Goto Program utilties
```
Name

thk123 replied:
Not sure exactly what this comment wants the name to change to? I've corrected the capitalisation if that's what you meant?

smowton replied:
Oh no, I thought you had copy pasted the header from a different file because the name didn't seem to correspond to the file name.

smowton commented on src/goto-programs/resolve_concrete_function_call.cpp:
```diff
@@ -0,0 +1,87 @@
+  const symbol_tablet &symbol_table):
+    symbol_table(symbol_table)
+{
+  class_hierarchy(symbol_table);
```
I'd recommend taking a const ref to the class hierarchy, as users are likely to have need of it themselves and it would be good to avoid rebuilding it unnecessarily.

thk123 replied:
I've provided both as options - in one case we have no use for the `class_hierarchy` but in the other 2 we already have one.

smowton commented on src/goto-programs/resolve_concrete_function_call.h:
```diff
@@ -0,0 +1,62 @@
+/*******************************************************************\
+
+ Module: Goto Program utilties
```
Name again

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,483 @@
+bool ci_lazy_methodst::operator()(
+  symbol_tablet &symbol_table,
+  lazy_methodst &lazy_methods,
+  std::function<void(
```
Add a typedef for this type in the header and use the typedef here.

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,483 @@
+  std::vector<const code_function_callt *> virtual_callsites;
 
-#include <string>
+  bool any_new_methods;
```
Should be initialized.

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,483 @@
+  do
+  {
+    any_new_methods=false;
+    while(!method_worklist2.empty())
```
I'm not a fan of this nested loop. Could we refactor this loop body into a method which returns whether there are any new methods?

thk123 replied:
Probably - but this isn't my code (just moved it) so I'd rather leave it for now. I can add a TODO comment or an issue if you think it is worth it?

reuk replied:
Sorry, didn't realise this wasn't yours (it's all green). I think it's fine to ignore for now.

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,483 @@
+          << " unreaclass_hierarchyable methods and globals"
+          << eom;
+
+  symbol_table.swap(keep_symbols);
```
Should this be a move assignment?

thk123 replied:
I don't know - since not my code would rather leave it unless you think its worth me finding out whether this is safe.

thk123 commented:
Resolved @smowton's concerns, @reuk could I get a re-review.

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,484 @@
+    needed_methods.push_back(self_method);
+  }
+
+  auto child_classes=class_hierarchy.get_children_trans(call_class);
```
Make const.

reuk commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -1,39 +1,484 @@
+    // on an opaque type (i.e. we don't have the class definition at this point)
+    // and will be created during the typecheck phase.
+    // We don't mark it as 'needed' as it doesn't exist yet to keep.
+    auto findit=symbol_table.symbols.find(to_symbol_expr(e).get_identifier());
```
Make const.

reuk commented on src/java_bytecode/ci_lazy_methods.h:
```diff
@@ -1,49 +1,107 @@
+    const irep_idt &component_method_name);
+
+  class_hierarchyt class_hierarchy;
+  const irep_idt main_class;
```
Having const-value members will suppress copy + move operations. Is this desired?

thk123 replied:
Since this is a functor class, I don't envisage it being copied or moved - should I explicitly remove the copy/move constructors?

reuk replied:
If it's an STL-compatible functor it *must* be copyable. If not, it might be helpful to explicitly disable the copy operations, and to supply an explanatory comment.

reuk commented on src/java_bytecode/ci_lazy_methods_needed.h:
```diff
@@ -0,0 +1,49 @@
+  symbol_table(_symbol_table)
+  {}
+
+  void add_needed_method(const irep_idt &);
```
These methods seem similar, it's weird they have different signatures.

thk123 replied:
You mean the fact `add_needed_class` returns a boolean? I agree - I will see if it is easy to add a sensible boolean to the `add_needed_method` (it is also a bit odd that one is stored in a set, and one is stored in a vector...)

thk123 replied:
I think this is non-trivial and since not my code would prefer to leave. It would either trivially return true (since it always adds something to the vector) or I could swap it to a set which _feels_ right, but would change the behaviour.

reuk replied:
Again, sorry about this - didn't realise it wasn't new code.

smowton replied:
(the difference was down to use case-- for classes a user needed to do insert-with-membership-test, for the methods it was sufficient to append to work list and dedup en masse)

reuk commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2728,39 +2729,48 @@ void java_bytecode_convert_method(
+    // methods with the default access modifier are only
+    // accessible within the same package.
+
+    else if(access==ID_default)
```
No `else` after `return`

thk123 replied:
I disagree - I think it makes it clearer we are essentially switching on the type of access

reuk replied:
This doesn't change the semantics of the code, so I'm happy for you to leave it, but the [LLVM standard](http://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return) has some nice examples of when the *no else after return* style makes sense.

thk123 replied:
That + the fact that all the cases are basic blocks ending in a return convinced me.

reuk commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -534,155 +255,30 @@ bool java_bytecode_languaget::do_ci_lazy_method_conversion(
-  symbol_tablet keep_symbols;
-
-  for(const auto &sym : symbol_table.symbols)
+  const auto &method_converter=[&](
```
This shouldn't be a reference.

reuk commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -258,32 +260,25 @@ void remove_virtual_functionst::get_functions(
+
+  resolve_concrete_function_callt get_virtual_call_target(
+    symbol_table, class_hierarchy);
+  const resolve_concrete_function_callt ::concrete_function_callt &
```
unnecessary space

reuk commented on src/goto-programs/resolve_concrete_function_call.h:
```diff
@@ -0,0 +1,64 @@
+        function_identifier(method_id)
+    {}
+
+    const irep_idt class_identifier;
```
Const members suppress copy and move operations. Is this desired?

thk123 commented:
@reuk Thanks for the careful review (and explanation of the const member variable impact) I've addressed all the comments apart from suggestions that were both a bit fiddly and not in my code (sorry should have made it clearer that most of this is moving code).

thk123 commented:
Pushed commits to hopefully fix the doxygen errors that caused Travis to fail.
